File: MDN/MDN.m
rand('twister',2);
randn('seed',2);
components = 3;
loss_function = MDN_Loss(components, 1);
layers = {};

dense = Dense(26, [1]);
dense.initialize();
layers{1} = dense;

tanh_layer = Tanh('tanh');
tanh_layer.set_input_shape(layers{1}.output_shape());
layers{2} = tanh_layer;

mdn_layer = MDN_Layer(components, [26], [1]);
mdn_layer.set_input_shape(layers{2}.output_shape());
mdn_layer.initialize();
layers{3} = mdn_layer;

n = 250;
d = 1;
y = rand(n,d);
noise = (0.2).*rand(n,d) - 0.1;
x = y + 0.3*sin(2*pi*y) + noise;
epochs = 1;
error = zeros(epochs, 1);

for i=1:epochs
    [y_pred, layers] = forward_run(x, layers);
    pii = y_pred{1};
    mu = y_pred{2};
    sigma = y_pred{3};
    loss = loss_function.loss(pii, sigma, mu, y);
    error(i) = loss;
    [dpi, dmu, dsigma] = loss_function.gradient(pii, sigma, mu, y);
    loss_grad = {};
    loss_grad{1} = dpi;
    loss_grad{2} = dmu;
    loss_grad{3} = dsigma;
    layers = backward_run(loss_grad, layers);
end
figure(2)
plot(error)

n_test = 250;
% y_test = rand(n_test,d);
% noise_test = (0.2).*rand(n_test,d) - 0.1;
% x_test = y_test + 0.3*sin(2*pi*y_test) + noise_test;
x_test = linspace(0,1,n_test);
x_test = transpose(x_test);
% [result,layers] = forward_run(x_test, layers);
% y_test = generate_ensemble(result);

function t = get_pi_idx(x, pdf)
    N = size(pdf,2);
    accumulate = 0;
    t = -1;
    for i=1:N
        accumulate = accumulate + pdf(i);
        if accumulate >= x
            t = i;
        end
    end
end

function result = generate_ensemble(y_test)
    M = 1;
    out_pi = y_test{1}; out_mu = y_test{2}; out_sigma = y_test{3};
    ntest = size(out_pi, 1);
    result = rand(ntest, M);
    rn = randn(ntest, M);
    mu = 0;
    std = 0;
    idx = 0;
    for j=1:M
        for i=1:ntest
            idx = get_pi_idx(result(i, j),out_pi(i,:));
            mu = out_mu(i, idx);
            std = out_sigma(i, idx);
            result(i, j) = mu + rn(i, j)*std;
        end
    end
end

function [output, layers] = forward_run(X, layers)
    output = X;
    for i=1:length(layers)
        curlayer = layers{i};
        output = curlayer.forward_pass(output);
    end
end

function layers = backward_run(loss_grad, layers)
    for i=length(layers):-1:1
        curlayer = layers{i};
        loss_grad = curlayer.backward_pass(loss_grad);
    end
end

File: MDN/MDN.py
import matplotlib.pyplot as plt
import numpy as np
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
import math

NSAMPLE = 2500
d = 1
t = 1
y_data = np.random.uniform(0, 1, (NSAMPLE, d)).astype(np.float32)
noise = np.random.uniform(-0.1, 0.1, (NSAMPLE, d)).astype(np.float32)
x_data = y_data + 0.3*np.sin(2*np.pi*y_data) + noise
x_test = np.linspace(0, 1, NSAMPLE).reshape(-1, 1).astype(np.float32)

plt.figure(figsize=(8, 8))
plot_out = plt.plot(x_data,y_data,'ro',alpha=0.3)
plt.show()

NHIDDEN = 24
STDEV = 0.5
KMIX = 24
NOUT = KMIX * 3

x = tf.placeholder(dtype=tf.float32, shape=[None,1], name="x")
y = tf.placeholder(dtype=tf.float32, shape=[None,1], name="y")

Wh = tf.Variable(tf.random_normal([1,NHIDDEN], stddev=STDEV, dtype=tf.float32))
bh = tf.Variable(tf.random_normal([1,NHIDDEN], stddev=STDEV, dtype=tf.float32))

Wo = tf.Variable(tf.random_normal([NHIDDEN,NOUT], stddev=STDEV, dtype=tf.float32))
bo = tf.Variable(tf.random_normal([1,NOUT], stddev=STDEV, dtype=tf.float32))

hidden_layer = tf.nn.tanh(tf.matmul(x, Wh) + bh)
output = tf.matmul(hidden_layer,Wo) + bo

def get_mixture_coef(output):
  out_pi = tf.placeholder(dtype=tf.float32, shape=[None,KMIX], name="mixparam")
  out_sigma = tf.placeholder(dtype=tf.float32, shape=[None,KMIX], name="mixparam")
  out_mu = tf.placeholder(dtype=tf.float32, shape=[None,KMIX], name="mixparam")

  out_pi, out_sigma, out_mu = tf.split(output, 3, 1)

  max_pi = tf.reduce_max(out_pi, 1, keep_dims=True)
  out_pi = tf.subtract(out_pi, max_pi)

  out_pi = tf.exp(out_pi)

  normalize_pi = tf.reciprocal(tf.reduce_sum(out_pi, 1, keep_dims=True))
  out_pi = tf.multiply(normalize_pi, out_pi)

  out_sigma = tf.exp(out_sigma)

  return out_pi, out_sigma, out_mu

out_pi, out_sigma, out_mu = get_mixture_coef(output)

oneDivSqrtTwoPI = 1 / math.sqrt(2*math.pi)
def tf_normal(y, mu, sigma):
  result = tf.subtract(y, mu)
  result = tf.multiply(result,tf.reciprocal(sigma))
  result = -tf.square(result)/2
  return tf.multiply(tf.exp(result),tf.reciprocal(sigma))*oneDivSqrtTwoPI

def get_lossfunc(out_pi, out_sigma, out_mu, y):
  result = tf_normal(y, out_mu, out_sigma)
  result = tf.multiply(result, out_pi)
  result = tf.reduce_sum(result, 1, keep_dims=True)
  result = -tf.log(result)
  return tf.reduce_mean(result)

lossfunc = get_lossfunc(out_pi, out_sigma, out_mu, y)
train_op = tf.train.AdamOptimizer().minimize(lossfunc)

sess = tf.InteractiveSession()
sess.run(tf.initialize_all_variables())

NEPOCH = 15000
loss = np.zeros(NEPOCH)
for i in range(NEPOCH):
  sess.run(train_op,feed_dict={x: x_data, y: y_data})
  loss[i] = sess.run(lossfunc, feed_dict={x: x_data, y: y_data})
  
NTEST = x_test.size
x_test = x_test.reshape(NTEST,1) # needs to be a matrix, not a vector

def get_pi_idx(x, pdf):
  N = pdf.size
  accumulate = 0
  for i in range(0, N):
    accumulate += pdf[i]
    if (accumulate >= x):
      return i
  print ('error with sampling ensemble')
  return -1

def generate_ensemble(out_pi, out_mu, out_sigma, M = 10):
  NTEST = x_test.size
  result = np.random.rand(NTEST, M)
  rn = np.random.randn(NTEST, M)
  mu = 0
  std = 0
  idx = 0

  for j in range(0, M):
    for i in range(0, NTEST):
      idx = get_pi_idx(result[i, j], out_pi[i])
      mu = out_mu[i, idx]
      std = out_sigma[i, idx]
      result[i, j] = mu + rn[i, j]*std
  return result

out_pi_test, out_sigma_test, out_mu_test = sess.run(get_mixture_coef(output), feed_dict={x: x_test})
y_test = generate_ensemble(out_pi_test, out_mu_test, out_sigma_test)
plt.figure(figsize=(8, 8))
plt.plot(x_data,y_data,'ro', x_test,y_test,'bo',alpha=0.3)
plt.show()